text/0000-dll-kind.md

```diff
@@ -0,0 +1,106 @@
+
+* If `#[link_ordinal]` is specified the idata section will map from the mangled symbol to the ordinal specified in the dll.
+* If `#[link_name]` is specified the idata section will map from the mangled symbol to the name specified in the dll, without any calling convention decorations added. If calling convention decorations are desired they must be specified explicitly in the value of the `#[link_name]` attribute.
+* If both `#[link_ordinal]` and `#[link_name]` are specified, an error will be emitted.
+* If neither `#[link_ordinal]` nor `#[link_name]` are specified, the idata section will map from the mangled symbol to its unmangled equivalent in the dll. The unmangled symbol will *not* have calling convention decorations.
+* If `#[no_mangle]` is specified an error will be emitted.
+
+The idata section that is produced is equivalent to the idata sections found in import libraries, and should result in identical code generation by the linker.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Additional complexity in the language through a new `kind` and a new attribute for specifying ordinals.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+The RFC as proposed would allow for full control over linking to symbols from dlls with syntax as close as possible to existing extern blocks.
+
+No alternatives are currently known other than the status quo.
+
+# Prior art
+[prior-art]: #prior-art
+
+Many non-native languages have the ability to import symbols from dlls, but this uses runtime loading by the language runtime and is not the same as what is being proposed here.
+
+Delphi is a native language that has the ability to import symbols from dlls without import libraries.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* Bikeshedding on attribute names.
+* Should this feature be extended to other platforms?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+With the features described in this RFC, we would be one step closer towards a fully standalone pure Rust target for Windows that does not rely on any external libraries (aside from the obvious and unavoidable runtime dependence on system libraries), allowing for easy installation and easy cross compilation.
+
+If that were to happen, we'd no longer need to pretend the pc-windows-gnu toolchain is standalone, and we'd be able to stop bundling MinGW bits entirely in favor of the user's own MinGW installation, thereby resolving a bunch of issues such as [rust-lang/rust#53454](https://github.com/rust-lang/rust/issues/53454).
+
+A future extension of this feature would be the ability to optionally lazily load such external functions, since Rust would naturally have all the information required to do so.
+
+Also users would stop complaining about having to install several gigabytes of VC++ just to link their Rust binaries.
```

Comment:
Per my previous comment, please add the following bits on extending this to non-Windows platforms:

Windows platforms need this feature most urgently, to avoid the need to generate and use import libraries, and we should prioritize implementation of this for Windows. However, other platforms would also benefit from the ability to build against a shared library without requiring the presence of that shared library at build time. For example, on Linux and other platforms using ELF shared libraries, the compiler would emit an ELF `NEEDED` entry for the specified shared library name, and an undefined symbol for each function declared. (On ELF platforms, using the `link_ordinal` attribute would produce an error.) On such platforms, the `link_name` attribute may also specify a symbol name that includes a symbol version, including the `@@`.